### ISSUE TITLE
One more revisit to CSV compatibility with Metadata

### ISSUE THREAD

**Issue by gkellogg** [w3c/csvw#534]
Recently, there’s been some discussion about alternative Metadata formats that may emerge in the future [[1]](https://lists.w3.org/Archives/Public/public-csv-wg/2015Apr/0090.html) [[2]](https://twitter.com/Gkellogg/status/594165661212901376). In fact, I think that other Metadata formats are perfectly viable, and can be used as an alternative way of generating the Annotated Tabular Data Model directly. As I result, I fear that placing processing instructions about the JSON version we define for our Metadata in the Data Model document might tie us too close to a particular format. Here are some things we might do to better separate these concerns:
- Describe different states for the Annotated Tabular Data Model, so that a version of the model prior to actually processing rows is defined; this should correspond to the result of processing the metadata file to create these annotations.
- Move Locating Metadata to the Metadata spec other than the embedded metadata portion, and reference that metadata standards MUST define a way of locating metadata resulting in the creation of the Data Model.
  - If no Metadata is found using any mechanism supported by the processor, default to the embedded metadata.
- A CSV is compatible with the Annotated Tabular Data Model based on comparing titles from the CSV with annotations, not metadata properties, thus moving the compatibility work out of the Metadata document and back into the Data Model document.
- Define the annotations in the Data Model using URLs rather than strings; we can continue to use these strings in the text, but in the preface, we can say that they are all interpreted as URLs relative to the CSVW namespace; this allows RDF toolchains to more easily create the Annotated Tabular Data Model, so that an RDF Graph might be more easily turned into (or used as the native representation for) the Annotated Tabular Data Model.
### Background

David Booth’s experiment [[1]](https://lists.w3.org/Archives/Public/public-csv-wg/2015Apr/0090.html) shows some novel use of metadata embedded in a CSV as defining the metadata (or ATDM) for processing other CSVs. With some work, a metadata file could probably created that would turn such a CSV into RDF which maps onto the ATDM.

Dan also provided an example [[2]](https://twitter.com/Gkellogg/status/594165661212901376) of using HTML to define the metadata, which I reduced to the following Turtle automatically:

```
@base <http://danbri.github.io/csvw-template/example.csv.html> .
@prefix csvw: <http://www.w3.org/ns/csvw#> .
@prefix dcterms: <http://purl.org/dc/terms/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfa: <http://www.w3.org/ns/rdfa#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<http://danbri.github.io/csvw-template/example.csv.html> rdfa:usesVocabulary csvw: .

<http://danbri.github.io/csvw-template/example.csv> a schema:DataDownload .

 [
    a csvw:Table, schema:Dataset;
    dcterms:creator "Dan Brickley";
    dcterms:modified "2014-05-09T15:44:58Z";
    dcterms:publisher "My Books";
    schema:creator "Dan Brickley";
    schema:dateModified "2014-05-09T15:44:58Z";
    schema:publisher "My Books";
    schema:url <http://danbri.github.io/csvw-template/example.csv>;
    csvw:tableSchema [
      a csvw:Schema;
      csvw:aboutUrl "http://librarything.com/isbn/{isbn}";
      csvw:columns [
        a csvw:Column;
        csvw:datatype "string";
        csvw:name "isbn";
        csvw:propertyUrl "http://purl.org/dc/terms/identifier";
        csvw:titles "ISBN-10";
        csvw:unique "true"
      ],  [
        a csvw:Column;
        csvw:datatype "string";
        csvw:name "title";
        csvw:propertyUrl "http://purl.org/dc/terms/title";
        csvw:titles "Book Title"
      ],  [
        a csvw:Column;
        csvw:datatype "string";
        csvw:name "author";
        csvw:propertyUrl "http://purl.org/dc/terms/creator";
        csvw:titles "Book Author"
      ];
      csvw:primaryKey "isbn"
    ];
    csvw:url <http://danbri.github.io/csvw-template/example.csv>
 ] .
```

This pretty much is a direct RDF serialization of an ATDM minus the row data, making a pretty compelling case for using arbitrary RDF serializations as a way to define an ATDM. Obviously, transforming a JSON Metadata file to Turtle would generate something quite similar:

```
@base <https://raw.githubusercontent.com/w3c/csvw/gh-pages/examples/tree-ops.csv-metadata.json> .
@prefix csvw: <http://www.w3.org/ns/csvw#> .
@prefix dc: <http://purl.org/dc/terms/> .
@prefix dcat: <http://www.w3.org/ns/dcat#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

 [
    dc:title "Tree Operations"@en;
    dc:license <http://opendefinition.org/licenses/cc-by/>;
    dc:modified "2010-12-31"^^xsd:date;
    dc:publisher [
      schema:name "Example Municipality"@en;
      schema:url <http://example.org>
    ];
    csvw:tableSchema [
      csvw:aboutUrl <https://raw.githubusercontent.com/w3c/csvw/gh-pages/examples/tree-ops.csv-metadata.json#gid-{GID}>;
      csvw:column ([
          dc:description "An identifier for the operation on a tree."@en;
          csvw:datatype xsd:string;
          csvw:name "GID";
          csvw:required true;
          csvw:title "GID"@en,
            "Generic Identifier"@en
        ] [
          dc:description "The street that the tree is on."@en;
          csvw:datatype xsd:string;
          csvw:name "on_street";
          csvw:title "On Street"@en
        ] [
          dc:description "The species of the tree."@en;
          csvw:datatype xsd:string;
          csvw:name "species";
          csvw:title "Species"@en
        ] [
          dc:description "The operation performed on the tree."@en;
          csvw:datatype xsd:string;
          csvw:name "trim_cycle";
          csvw:title "Trim Cycle"@en
        ] [
          dc:description "The date of the operation that was performed."@en;
          csvw:datatype [
            csvw:base "date";
            csvw:format "M/d/yyyy"
          ];
          csvw:name "inventory_date";
          csvw:title "Inventory Date"@en
        ]);
      csvw:primaryKey "GID"
    ];
    csvw:url "tree-ops.csv"^^xsd:anyURI;
    dcat:keyword "tree"@en,
      "street"@en,
      "maintenance"@en
 ] .
```

**Comment by iherman** [comment]
I am sorry @gkellogg, but I am not in favor of this change, for several reasons...
1. We are pretty late in this process to make a fairly radical change. It is radical in the sense of changing some of the emphasis and assumptions. We have just made an important change which was justified by simplifying the document and the implementations; this proposed change is different. If I look at our charter, our plans, etc, the proposed change would, with a high probability, lead to a delay. (Our plan is to issue a LCCR in June, i.e., which is in a month.) Of course, if the change is vital and essential, backed up by use cases, then we would have to consider it, but I do not think this is the case, see below.
2. We have agreed, right at the beginning of the work, that the community we are aiming at, i.e., publishers of CSV data, are not really familiar with RDF, nor do they want to invest into this. I.e., although defining metadata in Turtle is a technical possibility, I would still want to see convincing use cases for it before we took a major step to re-engineer the specifications to make this very explicit. (Beyond the fact that JSON-LD is, in fact, an RDF serialization, hence anyone can use Turtle for the metadata, if they so desire, and convert it to the JSON format.)
3. We have just had a discussion thread (see also issue #513) with people having reservation about the fact that the CSVW work is too dependent on RDF (which turned out to be a matter of a misconception based on an example). Introducing the concept of assigning URI-s to annotations would reinforce this view: the _only_ reason for doing so would be to accommodate RDF. Using URI-s in the annotated data model would be mostly incomprehensible to anyone else. We already do that in the metadata document, in fact, but we rely on JSON-LD to effectively hide this fact; this trick would not work well in the model document. I think that we would shoot in our own feet by doing that in the model document.
4. David's experiment is really interesting but, for me, falls under the reference to embedded metadata whose specification we explicitly pushed out of scope. I do not see any compelling reason why we would change that; if we did, we would have to go through the various use cases to see which form of embedded metadata is the most prevalent. 

Bottom line: as I said, I am not in favor of the change. Having said that, I do agree that the approach has merits: I would propose this to be listed as a potential future work for a future reincarnation of the Working Group.

(Actually, if we are talking about future work, I think looking more closely at the issue of embedded metadata and trying to come up with some standard for that is more important. I.e., going back to David's original example and see whether this could be worked out further with enough interest in the community.)

**Comment by gkellogg** [comment]
The proposal is not about RDF, it's about de-coupling the specifics of the JSON Metadata from the Data Model document. I don't think it's a radical change. That this might enable a future RDF-based Metadata format is a consequence of the change, not a specific goal.

**Comment by iherman** [comment]
I understand, but the motivation is _around_ RDF or embedded metadata. I have not seen any other reason for making a change at this point...

**Comment by JeniT** [comment]
I think we should keep it as it is. I think we're meeting the majority of use cases and I think we'd need really compelling evidence of demand for other metadata formats to introduce them. Whatever we do, expansion to support other metadata formats in the future would require a new version of the specs. So I'd rather cross that bridge when we come to it than worry trying to future proof this now.

**Comment by danbri** [comment]
I see the status quo as ok. 

There is a natural irony or tension here, and probably somewhat unavoidable given what we're attempting. 

The csv2rdf mechanisms we define actively encourage data to flow from classic CSV into the wider ecosystem, for mixing into SPARQL databases, JSON-LD and RDFa -oriented use cases, etc etc. However, anyone who actually does this is stepping beyond the W3C CSVW specs and in a sense moving out of compliance. If you used our specs to merge a mountain of annotated CSVs and republished the outputs as N-Triples or Turtle or as a SPARQL database, you could say that this was _facilitated by_ the W3C CSVW standards, but not that the _results_ were compliant with them. 

I suppose what we're saying, explicitly or otherwise, is that while RDF apps and developers might enjoy having lots of different formats, we want to be careful not to impose on CSVW processors an expectation that they'll need to handle these dozens of different RDF syntaxes. And so anyone publishing CSVW-based descriptions of tabular data SHOULD minimally make sure that those descriptions are accessible in our specified JSON-LD form, regardless of however else they might be shared.
